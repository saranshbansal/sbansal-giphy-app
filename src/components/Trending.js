import React, { Component } from "react";
import { fetchTrending } from "../resources/api";
import * as utils from "../resources/util";
import GifList from "./GifList";
import Loader from "./library/Loader";

class Trending extends Component {
  constructor(props) {
    super(props);

    this.refreshResults = this.refreshResults.bind(this);
    this.onScroll = this.onScroll.bind(this);
    this.onScroll = this.onScroll.bind(this);
  }

  state = {
    gifs: [],
    limit: 20,
    offset: 0,
    loading: false
  };

  componentDidMount() {
    this.refreshResults();

    window.addEventListener("scroll", this.onScroll);
  }

  componentWillUnmount() {
    window.removeEventListener("scroll", this.onScroll);
  }

  async refreshResults(offset = 0) {
    const { gifs, limit } = this.state;

    let gifsContent = gifs;

    try {
      const result = await fetchTrending(offset, limit);

      if (!!result && result.data) {
        result.data.map(gif => {
          return gifsContent.push({
            id: gif.id,
            imageUrl: gif.images.fixed_height_downsampled.url,
            staticImageUrl: gif.images.fixed_height_still.url,
            title: gif.title,
          });
        });
      }
    } catch (e) {
      console.error(e);
    } finally {
      this.setState({
        gifs: gifsContent,
        offset: offset,
        loading: false
      });
    }
  }

  onScroll() {
    const { offset, limit } = this.state;

    if (utils.scrolledToBottom()) {
      console.log("at the bottom");
      //fetch new data and append to existing list
      this.refreshResults(limit + offset + 1);
    }
  }

  render() {
    const { gifs, loading } = this.state;

    const hasResults = !!gifs && gifs.length > 0;
    const showResults = hasResults && !loading;

    return (
      <div className="trending">
        {loading && <Loader />}

        {showResults && (
          <div className="gifs">
            <GifList data={gifs} />
          </div>
        )}
      </div>
    );
  }
}

export default Trending;
