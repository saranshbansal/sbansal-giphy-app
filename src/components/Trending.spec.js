import React from "react";
import { shallow, mount } from "enzyme";
import Trending from "./Trending";

describe("Trending Gifs", () => {
  it("should render correctly.", () => {
    const wrapper = shallow(<Trending />);

    expect(wrapper).toMatchSnapshot();
  });

  it("should load a random GIFs on calling handleSearch fn.", async () => {
    const wrapper = mount(<Trending />);
    const instance = wrapper.instance();

    expect(wrapper.state("gifs").length).toBe(0);

    await instance.refreshResults();

    expect(wrapper.state("gifs").length).toBeGreaterThan(0);

  });
});
