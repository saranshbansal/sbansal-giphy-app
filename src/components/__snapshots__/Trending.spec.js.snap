// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Trending Gifs should render correctly. 1`] = `
<div
  className="trending"
/>
`;
